### PR TITLE
distributed: templated ring attention

### PR DESCRIPTION
**Pull request by d4l3k**
This adds a templated version of the ring attention forwards function as well as tests it with memory efficient attention. This doesn't add support for memory efficient attention in DTensor. That will be added in a follow up PR.

This templating is also a POC of how to support other attention ops such as Jagged/nested tensor and as well how to implement striped attention in a scalable way.

Misc changes:

* Fixes all_to_all_single autograd implementation with CUDA + adds NCCL test
* Adds compile support to the ring attention implementations (required some tweaks to process groups)

Test plan:

```
pytest test/distributed/_tensor/test_attention.py
pytest test/distributed/test_functional_api.py
```

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

### CODE REVIEW

File: torch/distributed/_tensor/experimental/attention.py [hunk 1]
```diff
@@ -54,6 +54,10 @@ def _merge_sdpa(
     """
     assert len(chunks) == len(logsumexps)
 
+    # LSE may be padded in the sequence dimension such as with memory efficient attention.
+    seq_len = chunks[0].size(2)
```
Comment (wanchaol) [inline torch/distributed/_tensor/experimental/attention.py:58]: is this a no-op in other attention op?

File: test/distributed/test_functional_api.py [hunk 2]
```diff
@@ -704,5 +705,65 @@ def run_with_backward():
         self.assertIsNotNone(t.grad)
 
 
+class TestFunctionalAutogradWithNCCL(MultiProcessTestCase):
```
Comment (wanchaol) [inline test/distributed/test_functional_api.py:708]: I think this should just inherit from `TestCollectivesWithNCCL` to unify the setups?

File: test/distributed/test_functional_api.py [hunk 3]
```diff
@@ -704,5 +705,65 @@ def run_with_backward():
...
+            sizes,
+            group
+        )
+
+        self.assertEqual(out.shape, t.shape)
+        self.assertIsNotNone(out.grad_fn)
+        self.assertTrue(out.requires_grad)
+        loss = out.sum()
+        loss.backward()
+        self.assertIsNotNone(t.grad)
```
Comment (wanchaol) [inline test/distributed/test_functional_api.py:765]: can we do a actual real assertion? i.e. use some non-random values for inputs so that we know the gradients expected and assert that.

**Review by wanchaol — COMMENTED** [review]
This looks reasonable, have a few suggestions and questions inlined

File: torch/distributed/_tensor/experimental/attention.py (hunk 1 above)
Comment (d4l3k) [reply torch/distributed/_tensor/experimental/attention.py:58]: yes exactly -- it's already size seq_len in flash attention. mem efficient attention the length is aligned to 32 len

File: test/distributed/test_functional_api.py (hunk 3 above)
Comment (d4l3k) [reply test/distributed/test_functional_api.py]: done

File: torch/distributed/_tensor/experimental/attention.py [hunk 4]
```diff
@@ -80,19 +84,151 @@ def _scaled_dot_product_ring_flash_attention(
...
+    op: AttentionOp,
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    *args: object,
+    is_causal: bool = False,
+    **kwargs: object,
+) -> Tuple[torch.Tensor, ...]:
+    """
+    This is a generalized ring attention implementation that can different multiple attention ops.
```
[the remark this answers is not in the document]
Comment (d4l3k) [reply torch/distributed/_tensor/experimental/attention.py]: done

File: torch/csrc/distributed/c10d/Functional.cpp [hunk 5]
```diff
@@ -6,6 +6,7 @@
 #include <torch/csrc/distributed/c10d/GroupRegistry.hpp>
 #include <torch/csrc/distributed/c10d/ProcessGroup.hpp>
 #include <torch/csrc/distributed/c10d/RankLocal.hpp>
+#include <iostream>
```
Comment (wanchaol) [inline torch/csrc/distributed/c10d/Functional.cpp:9]: this was debugging? need to remove this I think

**Review by wanchaol — APPROVED** [review]
lgtm! one more comment inlined

File: torch/csrc/distributed/c10d/Functional.cpp (hunk 5 above)
Comment (d4l3k) [reply torch/csrc/distributed/c10d/Functional.cpp:9]: sorry, good catch

**Comment by d4l3k** [comment]
@jithunnair-amd all_to_all_single is registered for the generic `Autograd` backend and not CUDA specific: https://github.com/pytorch/pytorch/blob/main/torch/csrc/distributed/c10d/Functional.cpp#L562

As far as I can tell there's nothing that's CUDA specific -- other than maybe we're setting the wrong device_type in the unit test? https://github.com/pytorch/pytorch/blob/main/test/distributed/_tensor/test_attention.py#L174-L176

**Comment by zhangvia** [comment]
@d4l3k it seems that _templated_ring_attention doesn't support attn_mask，right？

**Comment by XilunWu** [comment]
@zhangvia this is right, the only masking we currently support is causal masking. We expect to support flex attention soon which will support attn_mask.
